merkle/hashers/tree_hasher.go

```diff
@@ -33,6 +35,32 @@ type LogHasher interface {
 	Size() int
 }
 
+// InplaceLogHasher provides the hash functions needed to compute dense merkle
+// trees.
+// Implementations may hold state and are not expected to be thread
+// safe. They should not be shared between multiple callers. Where performance
+// is not critical the LogHasher interface provides a safer alternative.
+//
+// Note: Implementations are not expected to be thread safe. The caller must
+// ensure that concurrent calls are not made and is responsible for making
+// copies of slices where necessary to avoid data being overwritten.
+type InplaceLogHasher interface {
+	// EmptyRootInto supports returning a special case for the root of an empty tree.
+	// The passed in slice is mutated. The caller is responsible for making copies
+	// where necessary. Not thread safe.
+	EmptyRootInto([]byte) []byte
+	// HashLeafInto computes the hash of a leaf into an existing slice,
+	// which is mutated. The leaf and res slices may be the same. Not thread safe.
+	HashLeafInto(leaf, res []byte) ([]byte, error)
+	// HashChildrenInto computes interior nodes into an existing slice, which is
+	// mutated. The res slice may be the same as one of the l or r slices. Not
```

Comment:
ignore: I wonder what would happen if `res` was not equal to either `l` or `r`, but overlapped with one or both. I don't think it makes sense to clarify it in the comment though (no need to make this contract any stronger).

Reply:
Because the source is written to the hasher first before the result is truncated it should at least produce the right answer, not that this is a sensible thing to do.